### PR TITLE
go-sdk workflow quickstart

### LINKED ISSUE #965: [Workflow] Quickstart for Go SDK
Creating a Workflow sample in the Quickstarts based on this docs PR. https://github.com/dapr/docs/pull/3895

### PR DESCRIPTION
**Pull request by mikeee**
# Description

Adds a go workflow quickstart.

Tested locally, wip until the workflow PR (dapr/go-sdk#487) is merged and an RC is cut.

## Issue reference

Will close #965 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] The quickstart code compiles correctly
* [x] You've tested new builds of the quickstart if you changed quickstart code
* [x] You've updated the quickstart's README if necessary
* [x] If you have changed the steps for a quickstart be sure that you have updated the automated validation accordingly. All of our quickstarts have annotations that allow them to be executed automatically as code. For more information see [mechanical-markdown](https://github.com/dapr/mechanical-markdown#readme). For user guide with examples see [Examples](https://github.com/dapr/mechanical-markdown/tree/main/examples#mechanical-markdown-by-example).

### CODE REVIEW

**Comment by paulyuk** [comment]
@mikeee looks great so far.  Any chance you could set the target of this PR against the `release-1.13` branch instead of `master`?

**Comment by mikeee** [comment]
> @mikeee looks great so far.  Any chance you could set the target of this PR against the `release-1.13` branch base instead of `master`?  

Done ✅

**Comment by paulyuk** [comment]
Hey @mikeee, it looks like the test/check for Go is failing on this new quickstart.  It's pretty basic with a not found error:

```bash
	Actual stdout:
		ℹ️  Validating config and starting app "order-processor"
		ℹ️  Started Dapr with app id "order-processor". HTTP Port: 45753. gRPC Port: 33929
		ℹ️  Writing log files to directory : /home/runner/work/quickstarts/quickstarts/workflows/go/sdk/order-processor/.dapr/logs
		== APP - order-processor == go: cannot find main module, but found .git/config in /home/runner/work/quickstarts/quickstarts
		== APP - order-processor == 	to create a module there, run:
		== APP - order-processor == 	cd ../../../.. && go mod init
		The App process exited with error code: exit status 1
		
		ℹ️  Received signal to stop Dapr and app processes. Shutting down Dapr and app processes.
```

Could you please take a look?

One thing I usually do is run `make validate` in the same folder as the makefile on my local dev machine.  you can catch errors well before the CI build.

**Comment by mikeee** [comment]
> Hey @mikeee, it looks like the test/check for Go is failing on this new quickstart.  It's pretty basic with a not found error:
> 
> ```bash
> 	Actual stdout:
> 		ℹ️  Validating config and starting app "order-processor"
> 		ℹ️  Started Dapr with app id "order-processor". HTTP Port: 45753. gRPC Port: 33929
> 		ℹ️  Writing log files to directory : /home/runner/work/quickstarts/quickstarts/workflows/go/sdk/order-processor/.dapr/logs
> 		== APP - order-processor == go: cannot find main module, but found .git/config in /home/runner/work/quickstarts/quickstarts
> 		== APP - order-processor == 	to create a module there, run:
> 		== APP - order-processor == 	cd ../../../.. && go mod init
> 		The App process exited with error code: exit status 1
> 		
> 		ℹ️  Received signal to stop Dapr and app processes. Shutting down Dapr and app processes.
> ```
> 
> Could you please take a look?
> 
> One thing I usually do is run `make validate` in the same folder as the makefile on my local dev machine.  you can catch errors well before the CI build.

Yep, I'm aware of the error. The issue is that there is no release candidate cut for the go-sdk. I could test against `main`, is this suitable for quickstarts?
Leaving this PR as draft until a consensus is reached.

File: workflows/go/sdk/order-processor/main.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+			metadata, err := wfClient.FetchWorkflowMetadata(context.Background(), id)
+			if err != nil {
+				log.Fatalf("failed to fetch workflow: %v", err)
+			}
+			if totalCost > 50000 && !approvalSought && ((metadata.RuntimeStatus != workflow.StatusCompleted) || (metadata.RuntimeStatus != workflow.StatusFailed) || (metadata.RuntimeStatus != workflow.StatusTerminated)) {
+				approvalSought = true
+				promptForApproval(id)
+			}
+		}
+		// Sleep to not DoS the dapr dev instance
```
Comment (ItalyPaleAle) [inline workflows/go/sdk/order-processor/main.go:113]: Maybe tweak this comment so it doesn't sound like that dramatic?


```suggestion
		// Sleep before the next iteration
```

File: workflows/go/sdk/order-processor/workflow.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+			}
+			return OrderResult{Processed: false}, nil
+		}
+	}
+	if err := ctx.CallActivity(ProcessPaymentActivity, workflow.ActivityInput(PaymentRequest{
+		RequestID:          orderID,
+		ItemBeingPurchased: orderPayload.ItemName,
+		Amount:             orderPayload.TotalCost,
+		Quantity:           orderPayload.Quantity,
+	})).Await(nil); err != nil {
```
Comment (ItalyPaleAle) [inline workflows/go/sdk/order-processor/workflow.go:65]: This is a bit of a personal preference (and maybe a nit), but i find multi-line code within an `if` statement to be very hard to read


```suggestion
    err := ctx.
        CallActivity(ProcessPaymentActivity, workflow.ActivityInput(PaymentRequest{
            RequestID:          orderID,
            ItemBeingPurchased: orderPayload.ItemName,
            Amount:             orderPayload.TotalCost,
            Quantity:           orderPayload.Quantity,
        })).
        Await(nil)
	if err != nil {
```

File: workflows/go/sdk/order-processor/workflow.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+		if err := ctx.CallActivity(NotifyActivity, workflow.ActivityInput(Notification{Message: fmt.Sprintf("Order %s failed!", orderID)})).Await(nil); err != nil {
+			log.Printf("failed to notify of a failed order: %v", err)
+		}
+		return OrderResult{Processed: false}, err
+	}
+	if err := ctx.CallActivity(UpdateInventoryActivity, workflow.ActivityInput(PaymentRequest{RequestID: orderID, ItemBeingPurchased: orderPayload.ItemName, Amount: orderPayload.TotalCost, Quantity: orderPayload.Quantity})).Await(nil); err != nil {
```
Comment (ItalyPaleAle) [inline workflows/go/sdk/order-processor/workflow.go:71]: Please break this into its own for readability:


```suggestion
    err := ctx.
        CallActivity(UpdateInventoryActivity, workflow.ActivityInput(PaymentRequest{
            RequestID: orderID,
            ItemBeingPurchased: orderPayload.ItemName,
            Amount: orderPayload.TotalCost,
            Quantity: orderPayload.Quantity,
        })).
        Await(nil)
	if err != nil {
```

File: workflows/go/sdk/order-processor/workflow.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+		if err := ctx.CallActivity(NotifyActivity, workflow.ActivityInput(Notification{Message: fmt.Sprintf("Order %s failed!", orderID)})).Await(nil); err != nil {
+			log.Printf("failed to notify of a failed order: %v", err)
+		}
+		return OrderResult{Processed: false}, err
+	}
+
+	if err := ctx.CallActivity(NotifyActivity, workflow.ActivityInput(Notification{Message: fmt.Sprintf("Order %s has completed!", orderID)})).Await(nil); err != nil {
```
Comment (ItalyPaleAle) [inline workflows/go/sdk/order-processor/workflow.go:78]: Ditto

File: workflows/go/sdk/order-processor/workflow.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+)
+
+// OrderProcessingWorkflow is the main workflow for orchestrating activities in the order process.
+func OrderProcessingWorkflow(ctx *workflow.WorkflowContext) (any, error) {
+	orderID := ctx.InstanceID()
+	var orderPayload OrderPayload
+	if err := ctx.GetInput(&orderPayload); err != nil {
+		return nil, err
+	}
+	if err := ctx.CallActivity(NotifyActivity, workflow.ActivityInput(Notification{Message: fmt.Sprintf("Received order %s for %d %s - $%d", orderID, orderPayload.Quantity, orderPayload.ItemName, orderPayload.TotalCost)})).Await(nil); err != nil {
```
Comment (ItalyPaleAle) [inline workflows/go/sdk/order-processor/workflow.go:21]:
```suggestion
    err := ctx.
        CallActivity(NotifyActivity, workflow.ActivityInput(Notification{
            Message: fmt.Sprintf("Received order %s for %d %s - $%d", orderID, orderPayload.Quantity, orderPayload.ItemName, orderPayload.TotalCost),
        })).
        Await(nil)
	if err != nil {
```

**Comment by paulyuk** [comment]
> > Hey @mikeee, it looks like the test/check for Go is failing on this new quickstart.  It's pretty basic with a not found error:
> > ```shell
> > 	Actual stdout:
> > 		ℹ️  Validating config and starting app "order-processor"
> > 		ℹ️  Started Dapr with app id "order-processor". HTTP Port: 45753. gRPC Port: 33929
> > 		ℹ️  Writing log files to directory : /home/runner/work/quickstarts/quickstarts/workflows/go/sdk/order-processor/.dapr/logs
> > 		== APP - order-processor == go: cannot find main module, but found .git/config in /home/runner/work/quickstarts/quickstarts
> > 		== APP - order-processor == 	to create a module there, run:
> > 		== APP - order-processor == 	cd ../../../.. && go mod init
> > 		The App process exited with error code: exit status 1
> > 		
> > 		ℹ️  Received signal to stop Dapr and app processes. Shutting down Dapr and app processes.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you please take a look?
> > One thing I usually do is run `make validate` in the same folder as the makefile on my local dev machine.  you can catch errors well before the CI build.
> 
> Yep, I'm aware of the error. The issue is that there is no release candidate cut for the go-sdk. I have a module replace pointed at the branch with workflows merged and this quickstart is passing. I could test against `main`, is this suitable for quickstarts? Leaving this PR as draft until a consensus is reached.

Yes, if you can reference a go module built from go-sdk main branch that is fine.  My only requirement is the quickstart can clone/restore/run on my machine, and on the GH action machine without additional steps.

**Comment by paulyuk** [comment]
Hi @mikeee - let's figure out how to get that Validate Go check working.  It's still complaining about go.mod.

**Review by paulyuk — CHANGES_REQUESTED** [review]
need to get Go Validate check passing

File: workflows/go/sdk/order-processor/go.mod [hunk 6]
```diff
@@ -0,0 +1,29 @@
+module dapr_example
+
+go 1.21
+
+toolchain go1.21.6
```
Comment (paulyuk) [inline workflows/go/sdk/order-processor/go.mod:5]: Please just remove this line, which is unusual and breaks the test.  Line 3 with `go 1.21` directive is good enough to state desired version.
```suggestion
```

**Review by paulyuk — CHANGES_REQUESTED** [review]

File: workflows/go/sdk/order-processor/go.mod (hunk 6 above)
Comment (mikeee) [reply workflows/go/sdk/order-processor/go.mod:5]: I'd argue it's not unusual... on upgrading from <1.21 to any package that declares 1.21 will automatically include the toolchain declaration. I believe the reason it is passing for you is that you likely have 1.21 and the test action is currently using go1.18, there is nothing to suggest the tests would pass if this line is removed.

I included the fix in #986 that should address the root cause and get the test passing if it's picked into the 1.13 branch. I set go to 1.21 as this was erroneous but the toolchain directive should remain at parity with the SDK in my opinion.

Comment (mikeee) [reply workflows/go/sdk/order-processor/go.mod:5]: I've cherry-picked the change and included it in this PR. I believe 🤞 the go validation should now pass.

Comment (paulyuk) [reply workflows/go/sdk/order-processor/go.mod:5]: Ok good call @mikeee - thank you.

**Review by paulyuk — APPROVED** [review]
LGTM!

**Pull request merged**